pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -115,7 +116,8 @@
     private volatile long producerDeadline = 0; // gets set on first successful connection
 
     private final BatchMessageContainerBase batchMessageContainer;
-    private CompletableFuture<MessageId> lastSendFuture = CompletableFuture.completedFuture(null);
```

Comment:
I think it would be worth reverting the changes that remove that lastSendFuture field. 
The LastSendFutureWrapper field could co-exist with lastSendFuture field. In flushAsync, the code would first check if the future inside the lastSendFutureWrapper.lastSendFuture == lastSendFuture. If it's not, it would create a new lastSendFutureWrapper instance. This would reduce the overhead of the solution to minimum.

Reply:
Good idea. Great optimization. I have addressed done. PTAL again. :)